## 서울시 스마트 헬스케어 데이터 수집

 
[데이터 출처: 질병관리청-지역사회건강조사](https://chs.kdca.go.kr/chs/rawDta/rawDtaProvdMain.do)
<details>
    <summary><font color = '#81BEF7'>(클릭)>데이터 수집 과정 및 데이터 설명</font></summary>
    질병관리청에 자료를 요청하여 원시자료를 받았습니다.
    이 데이터를 이용해 대한민국 국민들의 음주량, 흡연량, 아침식사여부, 운동습관, BMI 등의 설문조사된 지표를
    바탕으로 체형을 분석하였습니다. 그중 서울시에 해당하는 사람들의 데이터만 추출해 분석하였습니다.
</details>
    

<details>    <summary><font size = 4><font color = '#FACC2E'>칼럼 설명을 보시려면 클릭하세요: </font></font></summary>

    'oba_02z1': '키': ## 해당 사람의 키(cm)
    'oba_03z1': '몸무게': ## 해당 사람의 체중(kg)
    'pha_04z1': '격렬한 신체활동 일수': ## 고강도 운동의 일수(일수/7일)
    'pha_07z1': '중증도 신체활동 일수', ## 중강도 운동의 일수(일수/7일)
    'phb_01z1': '걷기 실천 일수' ## 1주일간 걷기 운동 횟수(일수/7일)
    'drb_01z3': '연간 음주 빈도', ## 1년동안의 음주 빈도
                (거의 마시지 않음 = 1, 한달에 한 번 미만 = 2, 한 달에 1번 = 3, 한 달에 2~4번 = 4, 일주일에 2~3번 = 5, 일주일에 4번 이상 = 6, 응답거부 = 7, 모름 = 9)
    'drb_03z1': '한번 섭취시 음주량' ## 한 번 섭취 시 음주량(잔)
                (1~2잔 = 1, 3~4잔 = 2, 5~6잔 = 3, 7~9잔 = 4, 10잔이상 = 5, 응답거부 = 7, 비해당 = 8)
    'smb_01z1': '일반담배 하루 평균 흡연량': ## 하루 평균 흡연량(개비)
                (응답거부 = 777, 비해당 = 888, 모름 = 999)
    'nua_01z2': '주간 아침식사 일수', ## 최근 1년간 주당 아침식사 일수
                (주 5~7일 = 1, 주 3~4일 = 2, 주 1~2일 = 3, 주 0회 = 4, 응답거부 = 7, 모름 = 9)
    'oba_01z1': '체형(5단계)'
                (매우 마름 = 1, 약간 마름 = 2, 보통 = 3, 약간 비만 = 4, 매우 비만 = 5)
</details>


In [151]:
import pandas as pd
import numpy as np

#미리 받은 원시자료를 excel에서 서울시 데이터만 따로 분리하여 추출하였습니다.
obesity = pd.read_csv('chs22_서울만.csv')
obesity.head()

EXAMIN_YEAR     exmprs_no  age  sex  CTPRVN_CODE  PBHLTH_CODE  SPOT_NO  \
0         2022  2.022110e+24   71    2           11     11700017      181   
1         2022  2.022110e+24   74    1           11     11700017      181   
2         2022  2.022110e+24   86    1           11     11700017      181   
3         2022  2.022110e+24   81    2           11     11700017      181   
4         2022  2.022110e+24   39    2           11     11700017      181   

   HSHLD_CODE  MBHLD_CODE  DONG_TY_CODE  ...  sob_01z1  sob_02z1  soa_01z1  \
0           8           1             1  ...         4         3         2   
1           8           2             1  ...         3         3         2   
2          31           1             1  ...         7         1         2   
3          31           2             1  ...         3         3         2   
4          38           1             1  ...         4         1         1   

   soa_06z2  soa_07z1  sod_02z3  cog_01z1  cog_03z1  cog_05z1  coi_01z2  
0        88         8         1         3         2         2         3  
1        88         8         1         4         3         3         4  
2        88         8         1         1         1         1         3  
3        88         8         1         3         2         3         3  
4         5         1         5         3         3         2         3  

[5 rows x 161 columns]

### 분석에 필요한 칼럼만 추출

In [152]:
obesity_df = obesity[['oba_02z1', 'oba_03z1','drb_01z3', 'drb_03z1', 'smb_01z1', 
                      'pha_04z1', 'pha_07z1', 'phb_01z1', 'nua_01z2', 'oba_01z1']]
obesity_df

oba_02z1  oba_03z1  drb_01z3  drb_03z1  smb_01z1  pha_04z1  pha_07z1  \
0         152.0      48.0         8         8       888         0         0   
1         160.0      62.0         8         8        10         0         6   
2         170.0      57.0         1         8       888         0         0   
3         160.0      50.0         8         8       888         0         0   
4         156.0      56.0         6         4        15         0         0   
...         ...       ...       ...       ...       ...       ...       ...   
22938     162.0      70.0         1         8       888         0         0   
22939     174.0      73.0         8         8        10         0         0   
22940     161.0      48.0         6         5        20         0         0   
22941     165.0      50.0         1         8       888         0         0   
22942     170.0      48.0         1         8        10         0         0   

       phb_01z1  nua_01z2  oba_01z1  
0             4         1         3  
1             1         1         3  
2             0         4         3  
3             1         2         2  
4             1         2         3  
...         ...       ...       ...  
22938         0         1         3  
22939         5         1         2  
22940         5         4         3  
22941         5         1         3  
22942         5         1         1  

[22943 rows x 10 columns]

### 결측치 확인

In [153]:
obesity_df.isnull().sum()

oba_02z1    0
oba_03z1    0
drb_01z3    0
drb_03z1    0
smb_01z1    0
pha_04z1    0
pha_07z1    0
phb_01z1    0
nua_01z2    0
oba_01z1    0
dtype: int64

### 칼럼명을 이해하기 쉽게 변경

In [154]:
obesity_df = obesity_df.rename(columns = {
    'oba_02z1': '키', 
    'oba_03z1': '몸무게', 
    'pha_04z1': '격렬한 신체활동 일수', 
    'pha_07z1': '중증도 신체활동 일수', 
    'phb_01z1': '걷기 실천 일수',  
    'drb_01z3': '연간 음주 빈도', 
    'drb_03z1': '한번 섭취시 음주량', 
    'smb_01z1': '일반담배 하루 흡연량', 
    'nua_01z2': '주간 아침식사 일수',
    'oba_01z1': '체형(5단계)'
})
obesity_df


키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0      152.0  48.0         8           8          888            0   
1      160.0  62.0         8           8           10            0   
2      170.0  57.0         1           8          888            0   
3      160.0  50.0         8           8          888            0   
4      156.0  56.0         6           4           15            0   
...      ...   ...       ...         ...          ...          ...   
22938  162.0  70.0         1           8          888            0   
22939  174.0  73.0         8           8           10            0   
22940  161.0  48.0         6           5           20            0   
22941  165.0  50.0         1           8          888            0   
22942  170.0  48.0         1           8           10            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0                0         4           1        3  
1                6         1           1        3  
2                0         0           4        3  
3                0         1           2        2  
4                0         1           2        3  
...            ...       ...         ...      ...  
22938            0         0           1        3  
22939            0         5           1        2  
22940            0         5           4        3  
22941            0         5           1        3  
22942            0         5           1        1  

[22943 rows x 10 columns]

#### 변경한 Dataframe을 새로운 csv파일로 저장

In [155]:
obesity_df.to_csv('obesity_df.csv', encoding = 'cp949')

## 데이터 전처리

### 전체 데이터 분포 확인

In [156]:
obesity_df.describe()

키           몸무게      연간 음주 빈도    한번 섭취시 음주량   일반담배 하루 흡연량  \
count  22943.000000  22943.000000  22943.000000  22943.000000  22943.000000   
mean     549.120978    504.504239      3.999477      4.255764    778.749205   
std     6164.550828   6334.813193      2.289637      2.876616    289.174358   
min       61.000000     23.000000      1.000000      1.000000      1.000000   
25%      158.000000     55.000000      2.000000      1.000000    888.000000   
50%      164.000000     62.000000      4.000000      4.000000    888.000000   
75%      172.000000     72.000000      5.000000      8.000000    888.000000   
max    99999.000000  99999.000000      8.000000      8.000000    888.000000   

        격렬한 신체활동 일수   중증도 신체활동 일수      걷기 실천 일수    주간 아침식사 일수       체형(5단계)  
count  22943.000000  22943.000000  22943.000000  22943.000000  22943.000000  
mean       0.918145      1.192215      5.128754      2.021750      3.232097  
std        1.840993      2.171634      2.340861      1.309757      0.828221  
min        0.000000      0.000000      0.000000      1.000000      1.000000  
25%        0.000000      0.000000      4.000000      1.000000      3.000000  
50%        0.000000      0.000000      6.000000      1.000000      3.000000  
75%        1.000000      2.000000      7.000000      4.000000      4.000000  
max       99.000000     99.000000     99.000000      4.000000      7.000000

### 데이터 변수 타입 확인

In [157]:
obesity_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22943 entries, 0 to 22942
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   키            22943 non-null  float64
 1   몸무게          22943 non-null  float64
 2   연간 음주 빈도     22943 non-null  int64  
 3   한번 섭취시 음주량   22943 non-null  int64  
 4   일반담배 하루 흡연량  22943 non-null  int64  
 5   격렬한 신체활동 일수  22943 non-null  int64  
 6   중증도 신체활동 일수  22943 non-null  int64  
 7   걷기 실천 일수     22943 non-null  int64  
 8   주간 아침식사 일수   22943 non-null  int64  
 9   체형(5단계)      22943 non-null  int64  
dtypes: float64(2), int64(8)
memory usage: 1.8 MB


### 이상치 제거

#### '키' = 99999 인 행 제거

In [158]:
obesity_df[obesity_df['키'] == 99999]

키      몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
1065   99999.0  99999.0         1           8          888            0   
1071   99999.0     87.0         1           8          888            0   
1834   99999.0     56.0         8           8          888            0   
1844   99999.0     70.0         4           1          888            0   
1850   99999.0     58.0         8           8          888            0   
...        ...      ...       ...         ...          ...          ...   
21066  99999.0  99999.0         8           8          888            0   
21082  99999.0     62.0         1           8          888            0   
21086  99999.0     45.0         8           8          888            0   
21101  99999.0     60.0         8           8          888            0   
22819  99999.0     60.0         1           8          888            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
1065             0         0           1        4  
1071             0         6           1        3  
1834             0         7           4        2  
1844             0         7           1        2  
1850             0         6           1        3  
...            ...       ...         ...      ...  
21066            7         7           1        3  
21082            0         7           1        4  
21086            0         7           1        2  
21101            0         7           1        4  
22819            0         3           3        4  

[86 rows x 10 columns]

In [159]:
# '키' 칼럼의 값이 99999가 아닌 행들만 필터링하여 데이터프레임 업데이트
obesity_df1= obesity_df[obesity_df['키'] != 99999]
obesity_df1

키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0      152.0  48.0         8           8          888            0   
1      160.0  62.0         8           8           10            0   
2      170.0  57.0         1           8          888            0   
3      160.0  50.0         8           8          888            0   
4      156.0  56.0         6           4           15            0   
...      ...   ...       ...         ...          ...          ...   
22938  162.0  70.0         1           8          888            0   
22939  174.0  73.0         8           8           10            0   
22940  161.0  48.0         6           5           20            0   
22941  165.0  50.0         1           8          888            0   
22942  170.0  48.0         1           8           10            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0                0         4           1        3  
1                6         1           1        3  
2                0         0           4        3  
3                0         1           2        2  
4                0         1           2        3  
...            ...       ...         ...      ...  
22938            0         0           1        3  
22939            0         5           1        2  
22940            0         5           4        3  
22941            0         5           1        3  
22942            0         5           1        1  

[22857 rows x 10 columns]

In [160]:
#obesity_df1.describe()

#### '키' = 77777인 사람 제거

In [161]:
obesity_df1[obesity_df1['키'] == 77777]

키      몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
972   77777.0  77777.0         5           3          888            0   
1063  77777.0  77777.0         1           8          888            3   
1068  77777.0  77777.0         1           8          888            3   

      중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
972             0         6           1        4  
1063            3         7           1        4  
1068            3         6           1        3

In [162]:
obesity_df2 = obesity_df1[obesity_df1['키'] != 77777]
obesity_df2

키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0      152.0  48.0         8           8          888            0   
1      160.0  62.0         8           8           10            0   
2      170.0  57.0         1           8          888            0   
3      160.0  50.0         8           8          888            0   
4      156.0  56.0         6           4           15            0   
...      ...   ...       ...         ...          ...          ...   
22938  162.0  70.0         1           8          888            0   
22939  174.0  73.0         8           8           10            0   
22940  161.0  48.0         6           5           20            0   
22941  165.0  50.0         1           8          888            0   
22942  170.0  48.0         1           8           10            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0                0         4           1        3  
1                6         1           1        3  
2                0         0           4        3  
3                0         1           2        2  
4                0         1           2        3  
...            ...       ...         ...      ...  
22938            0         0           1        3  
22939            0         5           1        2  
22940            0         5           4        3  
22941            0         5           1        3  
22942            0         5           1        1  

[22854 rows x 10 columns]

In [163]:
#obesity_df2.describe()

#### 몸무게 >= 70000인 행 제거

In [164]:
obesity_df3 = obesity_df2[obesity_df2['몸무게'] < 70000]
obesity_df3

키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0      152.0  48.0         8           8          888            0   
1      160.0  62.0         8           8           10            0   
2      170.0  57.0         1           8          888            0   
3      160.0  50.0         8           8          888            0   
4      156.0  56.0         6           4           15            0   
...      ...   ...       ...         ...          ...          ...   
22938  162.0  70.0         1           8          888            0   
22939  174.0  73.0         8           8           10            0   
22940  161.0  48.0         6           5           20            0   
22941  165.0  50.0         1           8          888            0   
22942  170.0  48.0         1           8           10            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0                0         4           1        3  
1                6         1           1        3  
2                0         0           4        3  
3                0         1           2        2  
4                0         1           2        3  
...            ...       ...         ...      ...  
22938            0         0           1        3  
22939            0         5           1        2  
22940            0         5           4        3  
22941            0         5           1        3  
22942            0         5           1        1  

[22771 rows x 10 columns]

In [165]:
#obesity_df3.describe()

#### 연간 음주 빈도 중 '응답거부(7)' 에 해당하는 행 삭제

In [166]:
obesity_df3[obesity_df3['연간 음주 빈도'] == 7]

Empty DataFrame
Columns: [키, 몸무게, 연간 음주 빈도, 한번 섭취시 음주량, 일반담배 하루 흡연량, 격렬한 신체활동 일수, 중증도 신체활동 일수, 걷기 실천 일수, 주간 아침식사 일수, 체형(5단계)]
Index: []

해당 행 없으므로 제거 불필요

#### 연간 음주 빈도 중 '비해당(8)' 을 0으로 변경

In [167]:
obesity_df3.loc[obesity_df['연간 음주 빈도'] == 8, '연간 음주 빈도'] = 0
obesity_df3

C:\Users\82107\anaconda3\lib\site-packages\pandas\core\indexing.py:1817: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0      152.0  48.0         0           8          888            0   
1      160.0  62.0         0           8           10            0   
2      170.0  57.0         1           8          888            0   
3      160.0  50.0         0           8          888            0   
4      156.0  56.0         6           4           15            0   
...      ...   ...       ...         ...          ...          ...   
22938  162.0  70.0         1           8          888            0   
22939  174.0  73.0         0           8           10            0   
22940  161.0  48.0         6           5           20            0   
22941  165.0  50.0         1           8          888            0   
22942  170.0  48.0         1           8           10            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0                0         4           1        3  
1                6         1           1        3  
2                0         0           4        3  
3                0         1           2        2  
4                0         1           2        3  
...            ...       ...         ...      ...  
22938            0         0           1        3  
22939            0         5           1        2  
22940            0         5           4        3  
22941            0         5           1        3  
22942            0         5           1        1  

[22771 rows x 10 columns]

#### 한 번 섭취 음주량이 '응답거부(7)' , 혹은 '모름(9)' 인 행 제거

In [168]:
obesity_df3[obesity_df3['한번 섭취시 음주량'] == 7]

Empty DataFrame
Columns: [키, 몸무게, 연간 음주 빈도, 한번 섭취시 음주량, 일반담배 하루 흡연량, 격렬한 신체활동 일수, 중증도 신체활동 일수, 걷기 실천 일수, 주간 아침식사 일수, 체형(5단계)]
Index: []

In [169]:
obesity_df3[obesity_df3['한번 섭취시 음주량'] == 9]

Empty DataFrame
Columns: [키, 몸무게, 연간 음주 빈도, 한번 섭취시 음주량, 일반담배 하루 흡연량, 격렬한 신체활동 일수, 중증도 신체활동 일수, 걷기 실천 일수, 주간 아침식사 일수, 체형(5단계)]
Index: []

해당 행 없으므로 제거할 필요 없음

#### 한번 섭취시 음주량이 '비해당(8)' 인 행 0으로 변경

In [170]:
obesity_df3.loc[obesity_df['한번 섭취시 음주량'] == 8, '한번 섭취시 음주량'] = 0
obesity_df3

C:\Users\82107\anaconda3\lib\site-packages\pandas\core\indexing.py:1817: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0      152.0  48.0         0           0          888            0   
1      160.0  62.0         0           0           10            0   
2      170.0  57.0         1           0          888            0   
3      160.0  50.0         0           0          888            0   
4      156.0  56.0         6           4           15            0   
...      ...   ...       ...         ...          ...          ...   
22938  162.0  70.0         1           0          888            0   
22939  174.0  73.0         0           0           10            0   
22940  161.0  48.0         6           5           20            0   
22941  165.0  50.0         1           0          888            0   
22942  170.0  48.0         1           0           10            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0                0         4           1        3  
1                6         1           1        3  
2                0         0           4        3  
3                0         1           2        2  
4                0         1           2        3  
...            ...       ...         ...      ...  
22938            0         0           1        3  
22939            0         5           1        2  
22940            0         5           4        3  
22941            0         5           1        3  
22942            0         5           1        1  

[22771 rows x 10 columns]

In [171]:
#obesity_df3.describe()

#### 한번 섭취시 음주량이 '10잔 이상(5)' 인 사람들은 기준이 애매하므로 제거

In [172]:
obesity_df3[obesity_df3['한번 섭취시 음주량'] == 5]

키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
55     173.0  72.0         4           5          888            0   
61     169.0  70.0         5           5           10            0   
64     163.0  68.0         5           5          888            0   
73     161.0  70.0         5           5          888            1   
80     178.0  77.0         5           5          888            0   
...      ...   ...       ...         ...          ...          ...   
22898  163.0  73.0         5           5          888            0   
22910  175.0  68.0         4           5          888            5   
22932  168.0  85.0         2           5          888            5   
22933  169.0  75.0         6           5          888            2   
22940  161.0  48.0         6           5           20            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
55               0         0           1        4  
61               0         6           1        4  
64               2         5           4        4  
73               0         7           4        4  
80               3         2           1        3  
...            ...       ...         ...      ...  
22898            0         7           3        5  
22910            3         7           1        3  
22932            0         3           1        4  
22933            0         2           1        4  
22940            0         5           4        3  

[1615 rows x 10 columns]

In [173]:
obesity_df4 = obesity_df3[obesity_df3['한번 섭취시 음주량'] != 5]
obesity_df4

키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0      152.0  48.0         0           0          888            0   
1      160.0  62.0         0           0           10            0   
2      170.0  57.0         1           0          888            0   
3      160.0  50.0         0           0          888            0   
4      156.0  56.0         6           4           15            0   
...      ...   ...       ...         ...          ...          ...   
22937  167.0  63.0         1           0           10            7   
22938  162.0  70.0         1           0          888            0   
22939  174.0  73.0         0           0           10            0   
22941  165.0  50.0         1           0          888            0   
22942  170.0  48.0         1           0           10            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0                0         4           1        3  
1                6         1           1        3  
2                0         0           4        3  
3                0         1           2        2  
4                0         1           2        3  
...            ...       ...         ...      ...  
22937            7         5           4        3  
22938            0         0           1        3  
22939            0         5           1        2  
22941            0         5           1        3  
22942            0         5           1        1  

[21156 rows x 10 columns]

#### 하루 평균 흡연량이 '응답거부(7)'과 '모름(9)' 인 행 제거

In [174]:
obesity_df4[obesity_df4['일반담배 하루 흡연량'] == 777]

Empty DataFrame
Columns: [키, 몸무게, 연간 음주 빈도, 한번 섭취시 음주량, 일반담배 하루 흡연량, 격렬한 신체활동 일수, 중증도 신체활동 일수, 걷기 실천 일수, 주간 아침식사 일수, 체형(5단계)]
Index: []

In [175]:
obesity_df4[obesity_df4['일반담배 하루 흡연량'] == 999]

Empty DataFrame
Columns: [키, 몸무게, 연간 음주 빈도, 한번 섭취시 음주량, 일반담배 하루 흡연량, 격렬한 신체활동 일수, 중증도 신체활동 일수, 걷기 실천 일수, 주간 아침식사 일수, 체형(5단계)]
Index: []

해당 행 없으므로 제거 불필요

#### 하루 평균 흡연량이 '비해당(888)' 인 행 0으로 변경

In [176]:
obesity_df4.loc[obesity_df4['일반담배 하루 흡연량'] == 888, '일반담배 하루 흡연량'] = 0
obesity_df4

C:\Users\82107\anaconda3\lib\site-packages\pandas\core\indexing.py:1817: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0      152.0  48.0         0           0            0            0   
1      160.0  62.0         0           0           10            0   
2      170.0  57.0         1           0            0            0   
3      160.0  50.0         0           0            0            0   
4      156.0  56.0         6           4           15            0   
...      ...   ...       ...         ...          ...          ...   
22937  167.0  63.0         1           0           10            7   
22938  162.0  70.0         1           0            0            0   
22939  174.0  73.0         0           0           10            0   
22941  165.0  50.0         1           0            0            0   
22942  170.0  48.0         1           0           10            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0                0         4           1        3  
1                6         1           1        3  
2                0         0           4        3  
3                0         1           2        2  
4                0         1           2        3  
...            ...       ...         ...      ...  
22937            7         5           4        3  
22938            0         0           1        3  
22939            0         5           1        2  
22941            0         5           1        3  
22942            0         5           1        1  

[21156 rows x 10 columns]

In [177]:
#obesity_df4.describe()

####  격렬한 신체활동 일수가 '응답거부(77)' 와 '모름(99)' 인 행 제거

In [178]:
obesity_df4[obesity_df4['격렬한 신체활동 일수'] == 77]

Empty DataFrame
Columns: [키, 몸무게, 연간 음주 빈도, 한번 섭취시 음주량, 일반담배 하루 흡연량, 격렬한 신체활동 일수, 중증도 신체활동 일수, 걷기 실천 일수, 주간 아침식사 일수, 체형(5단계)]
Index: []

In [179]:
obesity_df4[obesity_df4['격렬한 신체활동 일수'] == 99]

키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
1255  160.5  78.0         4           1            0           99   

      중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
1255           99         5           4        4

In [180]:
obesity_df5 = obesity_df4[obesity_df4['격렬한 신체활동 일수'] != 99]
obesity_df5

키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0      152.0  48.0         0           0            0            0   
1      160.0  62.0         0           0           10            0   
2      170.0  57.0         1           0            0            0   
3      160.0  50.0         0           0            0            0   
4      156.0  56.0         6           4           15            0   
...      ...   ...       ...         ...          ...          ...   
22937  167.0  63.0         1           0           10            7   
22938  162.0  70.0         1           0            0            0   
22939  174.0  73.0         0           0           10            0   
22941  165.0  50.0         1           0            0            0   
22942  170.0  48.0         1           0           10            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0                0         4           1        3  
1                6         1           1        3  
2                0         0           4        3  
3                0         1           2        2  
4                0         1           2        3  
...            ...       ...         ...      ...  
22937            7         5           4        3  
22938            0         0           1        3  
22939            0         5           1        2  
22941            0         5           1        3  
22942            0         5           1        1  

[21155 rows x 10 columns]

#### 중증도 신체활동 일수가 '응답거부(77)' 와 '모름(99)' 인 행 제거

In [181]:
obesity_df5[obesity_df5['중증도 신체활동 일수'] == 77]

Empty DataFrame
Columns: [키, 몸무게, 연간 음주 빈도, 한번 섭취시 음주량, 일반담배 하루 흡연량, 격렬한 신체활동 일수, 중증도 신체활동 일수, 걷기 실천 일수, 주간 아침식사 일수, 체형(5단계)]
Index: []

In [182]:
obesity_df5[obesity_df5['중증도 신체활동 일수'] == 99]

키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
22936  155.0  43.0         0           0            0            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
22936           99        77           2        1

In [183]:
obesity_df6 = obesity_df5[obesity_df5['중증도 신체활동 일수'] != 99]
obesity_df6

키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0      152.0  48.0         0           0            0            0   
1      160.0  62.0         0           0           10            0   
2      170.0  57.0         1           0            0            0   
3      160.0  50.0         0           0            0            0   
4      156.0  56.0         6           4           15            0   
...      ...   ...       ...         ...          ...          ...   
22937  167.0  63.0         1           0           10            7   
22938  162.0  70.0         1           0            0            0   
22939  174.0  73.0         0           0           10            0   
22941  165.0  50.0         1           0            0            0   
22942  170.0  48.0         1           0           10            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0                0         4           1        3  
1                6         1           1        3  
2                0         0           4        3  
3                0         1           2        2  
4                0         1           2        3  
...            ...       ...         ...      ...  
22937            7         5           4        3  
22938            0         0           1        3  
22939            0         5           1        2  
22941            0         5           1        3  
22942            0         5           1        1  

[21154 rows x 10 columns]

In [184]:
obesity_df6.describe()

키           몸무게      연간 음주 빈도    한번 섭취시 음주량   일반담배 하루 흡연량  \
count  21154.000000  21154.000000  21154.000000  21154.000000  21154.000000   
mean     164.207994     63.412846      2.500189      1.341590      1.442895   
std        8.942226     12.325203      1.810120      1.343688      4.698550   
min       61.000000     23.000000      0.000000      0.000000      0.000000   
25%      158.000000     55.000000      1.000000      0.000000      0.000000   
50%      163.000000     62.000000      2.000000      1.000000      0.000000   
75%      170.000000     70.000000      4.000000      2.000000      0.000000   
max      198.000000    157.000000      6.000000      4.000000     80.000000   

        격렬한 신체활동 일수   중증도 신체활동 일수      걷기 실천 일수    주간 아침식사 일수       체형(5단계)  
count  21154.000000  21154.000000  21154.000000  21154.000000  21154.000000  
mean       0.872554      1.166352      5.110145      1.965869      3.217264  
std        1.689599      1.959424      2.204474      1.289975      0.821503  
min        0.000000      0.000000      0.000000      1.000000      1.000000  
25%        0.000000      0.000000      4.000000      1.000000      3.000000  
50%        0.000000      0.000000      6.000000      1.000000      3.000000  
75%        1.000000      2.000000      7.000000      3.000000      4.000000  
max        7.000000      7.000000      7.000000      4.000000      5.000000

### 스케일 맞추기(정규화)

주간 아침식사 일수의 범주가 다른 범주들과 반대로 라벨링이 되어 있으니 이에 대한 라벨링 재설정

(1 -> 3, 2 -> 2, 3 -> 1, 4 -> 0)

#### 주간 아침식사 일수의 라벨 재설정

In [185]:
# '주간 아침식사 일수' 칼럼의 라벨을 변경하기 위한 딕셔너리
new_labels = {
    1: 3,
    2: 2,
    3: 1,
    4: 0
}

# replace 메서드를 사용하여 칼럼의 값 매핑
obesity_df6['주간 아침식사 일수'] = obesity_df6['주간 아침식사 일수'].replace(new_labels)
obesity_df6

C:\Users\82107\AppData\Local\Temp/ipykernel_11844/2221585493.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  obesity_df6['주간 아침식사 일수'] = obesity_df6['주간 아침식사 일수'].replace(new_labels)


키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0      152.0  48.0         0           0            0            0   
1      160.0  62.0         0           0           10            0   
2      170.0  57.0         1           0            0            0   
3      160.0  50.0         0           0            0            0   
4      156.0  56.0         6           4           15            0   
...      ...   ...       ...         ...          ...          ...   
22937  167.0  63.0         1           0           10            7   
22938  162.0  70.0         1           0            0            0   
22939  174.0  73.0         0           0           10            0   
22941  165.0  50.0         1           0            0            0   
22942  170.0  48.0         1           0           10            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0                0         4           3        3  
1                6         1           3        3  
2                0         0           0        3  
3                0         1           2        2  
4                0         1           2        3  
...            ...       ...         ...      ...  
22937            7         5           0        3  
22938            0         0           3        3  
22939            0         5           3        2  
22941            0         5           3        3  
22942            0         5           3        1  

[21154 rows x 10 columns]

In [186]:
obesity_df6.describe()

키           몸무게      연간 음주 빈도    한번 섭취시 음주량   일반담배 하루 흡연량  \
count  21154.000000  21154.000000  21154.000000  21154.000000  21154.000000   
mean     164.207994     63.412846      2.500189      1.341590      1.442895   
std        8.942226     12.325203      1.810120      1.343688      4.698550   
min       61.000000     23.000000      0.000000      0.000000      0.000000   
25%      158.000000     55.000000      1.000000      0.000000      0.000000   
50%      163.000000     62.000000      2.000000      1.000000      0.000000   
75%      170.000000     70.000000      4.000000      2.000000      0.000000   
max      198.000000    157.000000      6.000000      4.000000     80.000000   

        격렬한 신체활동 일수   중증도 신체활동 일수      걷기 실천 일수    주간 아침식사 일수       체형(5단계)  
count  21154.000000  21154.000000  21154.000000  21154.000000  21154.000000  
mean       0.872554      1.166352      5.110145      2.034131      3.217264  
std        1.689599      1.959424      2.204474      1.289975      0.821503  
min        0.000000      0.000000      0.000000      0.000000      1.000000  
25%        0.000000      0.000000      4.000000      1.000000      3.000000  
50%        0.000000      0.000000      6.000000      3.000000      3.000000  
75%        1.000000      2.000000      7.000000      3.000000      4.000000  
max        7.000000      7.000000      7.000000      3.000000      5.000000

In [187]:
# obesity_df6을 obesity_df_prep라는 새로운 변수에 저장
obesity_df_prep = obesity_df6
obesity_df_prep

키   몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0      152.0  48.0         0           0            0            0   
1      160.0  62.0         0           0           10            0   
2      170.0  57.0         1           0            0            0   
3      160.0  50.0         0           0            0            0   
4      156.0  56.0         6           4           15            0   
...      ...   ...       ...         ...          ...          ...   
22937  167.0  63.0         1           0           10            7   
22938  162.0  70.0         1           0            0            0   
22939  174.0  73.0         0           0           10            0   
22941  165.0  50.0         1           0            0            0   
22942  170.0  48.0         1           0           10            0   

       중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0                0         4           3        3  
1                6         1           3        3  
2                0         0           0        3  
3                0         1           2        2  
4                0         1           2        3  
...            ...       ...         ...      ...  
22937            7         5           0        3  
22938            0         0           3        3  
22939            0         5           3        2  
22941            0         5           3        3  
22942            0         5           3        1  

[21154 rows x 10 columns]

In [188]:
obesity_df_prep.to_csv('obesity_df_prep.csv', encoding = 'cp949')

#### 각 칼럼들에 대하여 표준 정규화(z-score 이용)
$$ z = \frac{(X - m)}{\sigma} $$


In [193]:
from sklearn.preprocessing import StandardScaler

# StandardScaler 객체 생성
scaler = StandardScaler()

# fit_transform 메서드를 사용하여 표준 정규화 수행
obesity_df_prep_std = pd.DataFrame(scaler.fit_transform(obesity_df_prep.iloc[:,:-1]), columns=obesity_df_prep.columns[:-1])
obesity_df_prep_std[obesity_df_prep.columns[-1]] = obesity_df_prep[obesity_df_prep.columns[-1]].values
obesity_df_prep_std.head()

키       몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0 -1.365240 -1.250544 -1.381261   -0.998463    -0.307101    -0.516439   
1 -0.470587 -0.114633 -1.381261   -0.998463     1.821266    -0.516439   
2  0.647729 -0.520316 -0.828799   -0.998463    -0.307101    -0.516439   
3 -0.470587 -1.088271 -1.381261   -0.998463    -0.307101    -0.516439   
4 -0.917913 -0.601452  1.933515    1.978490     2.885449    -0.516439   

   중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0    -0.595266 -0.503599    0.748768        3  
1     2.466930 -1.864500    0.748768        3  
2    -0.595266 -2.318134   -1.576913        3  
3    -0.595266 -1.864500   -0.026459        2  
4    -0.595266 -1.864500   -0.026459        3

In [192]:
obesity_df_prep_std.to_csv('obesity_std.csv', encoding = 'cp949')

#### MinMax-Scaling을 이용한 정규화
$$ X_s = \frac{x - min}{\max-min} $$

In [194]:
from sklearn.preprocessing import MinMaxScaler

mm_scaler = MinMaxScaler()

obesity_df_prep_mm = pd.DataFrame(mm_scaler.fit_transform(obesity_df_prep.iloc[:, :-1]), columns = obesity_df_prep.columns[:-1])
obesity_df_prep_mm[obesity_df_prep.columns[-1]] = obesity_df_prep[obesity_df_prep.columns[-1]].values
obesity_df_prep_mm.head()

키       몸무게  연간 음주 빈도  한번 섭취시 음주량  일반담배 하루 흡연량  격렬한 신체활동 일수  \
0  0.664234  0.186567  0.000000         0.0       0.0000          0.0   
1  0.722628  0.291045  0.000000         0.0       0.1250          0.0   
2  0.795620  0.253731  0.166667         0.0       0.0000          0.0   
3  0.722628  0.201493  0.000000         0.0       0.0000          0.0   
4  0.693431  0.246269  1.000000         1.0       0.1875          0.0   

   중증도 신체활동 일수  걷기 실천 일수  주간 아침식사 일수  체형(5단계)  
0     0.000000  0.571429    1.000000        3  
1     0.857143  0.142857    1.000000        3  
2     0.000000  0.000000    0.000000        3  
3     0.000000  0.142857    0.666667        2  
4     0.000000  0.142857    0.666667        3

In [196]:
obesity_df_prep_mm.to_csv('obesity_minmax.csv', encoding = 'cp949')

<details>
    <summary><font size = 4><font color = '#FA5858'>(클릭)전처리 결과</font></font></summary>
    
    전처리 결과, 총 22943개의 행 중 1789개의 행이 제거되거나 값이 변경되어, 현재 총 21154개의 행만이 남아있다.
    
    #전처리한 행과 열들은 다음과 같다.
    
    1. '키': 이상치 행 제거(77777 or 99999)
    2. '몸무게': 이상치 행 제거(77777 or 99999)
    3. '연간 음주 빈도': 응답거부(7) 제거
    4. '연간 음주 빈도': 비해당(8) 값 0으로 변경
    5. '한번 섭취시 음주량': 비해당(8) 값 0으로 변경
    6. '일반담배 하루 흡연량': 비해당(888)값 0으로 변경
    7. '격렬한 신체활동 일수': 모름(99)인 행 제거
    8. '중증도 신체활동 일수': 모름(99)인 행 제거
    9. '주간 아침식사 일수': 스케일 변경(1 --> 3, 2 --> 2, 3 --> 1, 4 --> 0)
    
    별도의 csv 파일(obesity_df_prep.csv)로 저장
    
    #표준 정규화 처리:
    마지막 칼럼인 '체형(5단계)'를 제외한 나머지 칼럼들에 대하여
    sklearn.preprocessing 내의 StandardScaler 이용 (평균 = 0, 표준편차 = 1 로 변환)
    별도의 csv 파일(obesity_std.csv)로 저장
    
    #MinMax 정규화 처리:
    마지막 칼럼인 '체형(5단계)'를 제외한 나머지 칼럼들에 대하여
    sklearn.preprocessing 내의 MinMaxScaler 이용 (최솟값부터 최댓값까지 0과 1사이 값으로 변환)
    별도의 csv 파일(obesity_minmax.csv)로 저장
    
</details>